### PR TITLE
Lecture2 Task preliminary implementation

### PR DESCRIPTION
**Pull request by carmenhorendum**

### CODE REVIEW

File: Lecture2/Task2/Student.java [hunk 1]
```diff
@@ -0,0 +1,48 @@
+public class Student implements Comparable<Student> {
+
+	private int id, age;
```
Comment (nemec8g) [inline Lecture2/Task2/Student.java]: all field - final, as it init only in constructor

File: Lecture2/Task2/TaskTwoMain.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
+import java.util.*;
+
+public class TaskTwoMain {
+	public static void main(String[] args) {
+		final String FIRST_LETTER = "M";
+		List<Student> students = new ArrayList<>();
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: create factory class to generate list of students 
`StudentFactory.getStudents(int amount)`

File: Lecture2/Task2/TaskTwoMain.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
+import java.util.*;
+
+public class TaskTwoMain {
+	public static void main(String[] args) {
+		final String FIRST_LETTER = "M";
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: move constant from main method

File: Lecture2/Task2/TaskTwoMain.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+		Map<Integer, Student> studentMapAboveThree = new HashMap<>();
+		for (Student key : studentsMap.values()) {
+			if (key.getId() > 3) {
+				studentMapAboveThree.put(key.getId(), key);
+				System.out.println(key.getName());
+			}
+		}
+	}
+
+	private static List<Student> getStudentsM(List<Student> students, String FIRST_LETTER) {
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: Create `Filter` class for such method to get some students

File: Lecture2/Task2/TaskTwoMain.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+			System.out.println(student.getAge() + " " + student.getName());
+		}
+
+		System.out.println(String.format("First letter of the name is %s:", FIRST_LETTER));
+		List<Student> studentsM = getStudentsM(students, FIRST_LETTER);
+		for (Student student : studentsM) {
+			System.out.println(student.getName());
+		}
+
+		double average = 0;
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: no needed local var

File: Lecture2/Task3/TaskThreeMain.java [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+		students.add(new Student(1, 60, "Rudolf", "Schenker"));
+		students.add(new Student(2, 75, "Klaus", "Meine"));
+		students.add(new Student(3, 55, "Matthias", "Jabs"));
+		students.add(new Student(4, 58, "Paweł", "Mąciwoda"));
+		students.add(new Student(5, 64, "Mikkey", "Dee"));
+
+		System.out.println("List sorted by Age:");
+		students = getSortByAge(students);
+		for (Student student : students) {
```
Comment (nemec8g) [inline Lecture2/Task3/TaskThreeMain.java]: use streamapi

File: Lecture2/Task3/TaskThreeMain.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+			System.out.println(student.getAge() + " " + student.getName());
+		}
+
+		System.out.println(String.format("First letter of the name is %s:", FIRST_LETTER));
+		List<Student> studentsM = getStudentsM(students, FIRST_LETTER);
+		for (Student student : studentsM) {
```
Comment (nemec8g) [inline Lecture2/Task3/TaskThreeMain.java]: use streamapi

File: Lecture2/Task3/TaskThreeMain.java (hunk 6 above)
Comment (carmenhorendum) [reply Lecture2/Task3/TaskThreeMain.java]: done

File: Lecture2/Task3/TaskThreeMain.java (hunk 7 above)
Comment (carmenhorendum) [reply Lecture2/Task3/TaskThreeMain.java]: done

File: Lecture2/Task2/TaskTwoMain.java (hunk 3 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: done

File: Lecture2/Task2/Student.java (hunk 1 above)
Comment (carmenhorendum) [reply Lecture2/Task2/Student.java]: done

File: Lecture2/Task2/TaskTwoMain.java (hunk 4 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: done

File: Lecture2/Task2/TaskTwoMain.java (hunk 5 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: fixed

File: Lecture2/Task2/TaskTwoMain.java (hunk 2 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: Will be implemented in the scope of Lecture 3 Hometask

File: Lecture2/Task1/builder/Guitar.java [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+
+import product.Brand;
+
+public class Guitar {
+	private Brand brand;
+	private String color;
+	private int stringsCount;
+
+	public Guitar() {
+		this.brand = brand;
```
Comment (nemec8g) [inline Lecture2/Task1/builder/Guitar.java]: brand is null 
this expression doesn't have any sense
you should pass all required param in constructor
public Guitar(Brand brand, String color) {
this.brand = brand
etc
}

File: Lecture2/Task1/builder/GuitarBuilder.java [hunk 9]
```diff
@@ -0,0 +1,27 @@
+package builder;
+
+import product.Brand;
+
+public class GuitarBuilder {
```
Comment (nemec8g) [inline Lecture2/Task1/builder/GuitarBuilder.java]: what is the responsibility of this class? 
you have all this method in Guitar class - use it instead

File: Lecture2/Task1/builder/GuitarBuilderDirector.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package builder;
+
+import product.Brand;
+
+public class GuitarBuilderDirector {
```
Comment (nemec8g) [inline Lecture2/Task1/builder/GuitarBuilderDirector.java]: This class looks line GuitarFactory

File: Lecture2/Task1/builder/Guitar.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
+package builder;
+
+import product.Brand;
+
+public class Guitar {
```
Comment (nemec8g) [inline Lecture2/Task1/builder/Guitar.java]: you have to create hierarchy of guitar

Create to class acustic and electric guitar which extents from guitar

File: Lecture2/Task2/TaskTwoMain.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+		students.add(new Student(5, 64, "Mikkey", "Dee"));
+
+		Collections.sort(students);
+		System.out.println("List sorted by Age:");
+		for (Student student : students) {
+			System.out.println(student.getAge() + " " + student.getName());
+		}
+
+		System.out.println(String.format("First letter of the name is %s:", FIRST_LETTER));
+		Filter.getStudentsM(students, FIRST_LETTER);
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: method `getStudents`  should return list of students

File: Lecture2/Task2/TaskTwoMain.java [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+
+		System.out.println(String.format("Average age of students is:\n%s",
+				getAvarageAge(average, students)));
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: what for you pass `average`  in method? List of students is enough for this method

File: Lecture2/Task2/TaskTwoMain.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+		Map<Integer, Student> studentMapAboveThree = new HashMap<>();
+		for (Student key : studentsMap.values()) {
+			if (key.getId() > 3) {
+				studentMapAboveThree.put(key.getId(), key);
+				System.out.println(key.getName());
+			}
+		}
+	}
+
+	private static double getAvarageAge(double average, List<Student> students) {
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: please create one common interface with methods `getAvarageAge ` , `getStudentsListToMap `  and other
create two implementation for stream/non stream 

and use this interface in task 3 and 4

File: Lecture2/Task2/TaskTwoMain.java [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+		}
+
+		System.out.println(String.format("First letter of the name is %s:", FIRST_LETTER));
+		Filter.getStudentsM(students, FIRST_LETTER);
+
+		System.out.println(String.format("Average age of students is:\n%s",
+				getAvarageAge(average, students)));
+
+		System.out.println("Names from the Map:");
+		Map<Integer, Student> studentsMap = getStudentsListToMap(students);
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: move all this methods to interface

File: Lecture2/Task2/TaskTwoMain.java [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+		for (Student key : studentsMap.values()) {
+			System.out.println(key.getName());
+		}
+
+		System.out.println("Names from the Map (id > 3):");
+		Map<Integer, Student> studentMapAboveThree = new HashMap<>();
+		for (Student key : studentsMap.values()) {
```
Comment (nemec8g) [inline Lecture2/Task2/TaskTwoMain.java]: Student student :

File: Lecture2/Task3/TaskThreeMain.java [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+		System.out.println(String.format("First letter of the name is %s:", FIRST_LETTER));
+		List<Student> studentsM = getStudentsM(students, FIRST_LETTER);
+		studentsM.forEach(System.out::println);
+
+		System.out.println(String.format("Average age of students is:\n%s",
+				getAverageAge(students)));
+
+		System.out.println("Names from the Map:");
+		Map<Integer, Student> studentsMap;
+        	getStudentsListToMap(students);
```
Comment (nemec8g) [inline Lecture2/Task3/TaskThreeMain.java:32]: you ignore getStudentsListToMap result and print initial list of students

File: Lecture2/Task4/TaskFourMain.java [hunk 18]
```diff
@@ -0,0 +1,58 @@
...
+			list.remove(random.nextInt(10));
+		}
+		return System.currentTimeMillis() - startTime;
+	}
+
+	private static long getFindElementTime(int CHANGE_SIZE, List<Integer> list) {
+		long startTime = System.currentTimeMillis();
+		Random random = new Random();
+		for (int i = 0; i < CHANGE_SIZE; i++) {
+			list.indexOf(random.nextInt(10));
```
Comment (nemec8g) [inline Lecture2/Task4/TaskFourMain.java:54]: you have to find item by valuе, not index

File: Lecture2/Task4/TaskFourMain.java [hunk 19]
```diff
@@ -0,0 +1,58 @@
...
+			list.add(random.nextInt(10));
+		}
+		return System.currentTimeMillis() - startTime;
+	}
+
+	private static long getDeleteElementTime(int CHANGE_SIZE, List<Integer> list) {
+		long startTime = System.currentTimeMillis();
+		Random random = new Random();
+		for (int i = 0; i < CHANGE_SIZE; i++) {
+			list.remove(random.nextInt(10));
```
Comment (nemec8g) [inline Lecture2/Task4/TaskFourMain.java:45]: you have to remove item by valuе, not index

File: Lecture2/Task1/builder/Guitar.java (hunk 11 above)
Comment (carmenhorendum) [reply Lecture2/Task1/builder/Guitar.java]: done

File: Lecture2/Task2/TaskTwoMain.java (hunk 12 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: fixed

File: Lecture2/Task2/TaskTwoMain.java (hunk 13 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: fixed

File: Lecture2/Task2/TaskTwoMain.java (hunk 14 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: done for 2

File: Lecture2/Task2/TaskTwoMain.java (hunk 15 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: done

File: Lecture2/Task2/TaskTwoMain.java (hunk 16 above)
Comment (carmenhorendum) [reply Lecture2/Task2/TaskTwoMain.java]: fixed

File: Lecture2/Task3/TaskThreeMain.java (hunk 17 above)
Comment (carmenhorendum) [reply Lecture2/Task3/TaskThreeMain.java:32]: fixed

File: Lecture2/Task4/TaskFourMain.java (hunk 18 above)
Comment (carmenhorendum) [reply Lecture2/Task4/TaskFourMain.java:54]: not to be fixed

File: Lecture2/Task4/TaskFourMain.java (hunk 19 above)
Comment (carmenhorendum) [reply Lecture2/Task4/TaskFourMain.java:45]: not to be fixed

File: Lecture2/Task1/builder/Guitar.java (hunk 8 above)
Comment (carmenhorendum) [reply Lecture2/Task1/builder/Guitar.java]: fixed

File: Lecture2/Task1/builder/GuitarBuilder.java (hunk 9 above)
Comment (carmenhorendum) [reply Lecture2/Task1/builder/GuitarBuilder.java]: discussed

File: Lecture2/Task1/builder/GuitarBuilderDirector.java (hunk 10 above)
Comment (carmenhorendum) [reply Lecture2/Task1/builder/GuitarBuilderDirector.java]: discussed

**Pull request merged**